### PR TITLE
[dtensor] add backward support for scaled dot product attention (flash-attention)

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122569
* __->__ #122541

As titled, as a followup to the forward part #120298.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_tensor/ops/matrix_ops.py [hunk 1]
```diff
@@ -224,3 +224,89 @@ def scaled_dot_product_attention_strategy(
...
+    q_input_strategy = op_schema.args_schema[1]
+    assert isinstance(q_input_strategy, OpStrategy)
+    # assuming q/k/v have the same shape
+    qkv_shape = q_input_strategy.output_shape
+
+    # there might be extra tensor inputs
+    # cum_seq_q, cum_seq_k, philox_seed, philox_offset
+    # at indices 6, 7, 12, 13, respectively
+    num_extra_inputs = sum(
+        [strategy is not None for strategy in op_schema.args_schema[6:14]]
```
Comment (Skylion007) [inline torch/distributed/_tensor/ops/matrix_ops.py:242]:
```suggestion
        (strategy is not None for strategy in op_schema.args_schema[6:14])
```

File: torch/distributed/_tensor/ops/matrix_ops.py [hunk 2]
```diff
@@ -224,3 +224,93 @@ def scaled_dot_product_attention_strategy(
...
+
+        single_mesh_dim_strategies.append(all_replicate)
+
+        # second we can accept the sharding pattern of tensor parallelism, which
+        # shard on the num of head dim
+        grad_output_sharding = Shard(1)
+        qkv_sharding = Shard(1)  # num head dim
+        output_sharding = Shard(1)  # num head dim
+        logsumexp_sharding = Shard(1)  # num head dim
+        grad_qkv_sharding = Shard(1)  # num head dim
```
Comment (XilunWu) [inline torch/distributed/_tensor/ops/matrix_ops.py:263]: Is this the only case we plan to support for now? Will we need to support more cases?

**Review by XilunWu — COMMENTED** [review]
overall LGTM. I can stamp the PR if you think this is ready.

File: torch/distributed/_tensor/ops/matrix_ops.py (hunk 2 above)
Comment (tianyu-l) [reply torch/distributed/_tensor/ops/matrix_ops.py:263]: AFAIK, currently TP/SP is the only use case so we only need to support head_dim sharding for now. We can always support other cases later if needed. Besides, Wanchao mentioned in the comment of the forward part:

"NOTE: currently we only support some simple strategies to support tensor parallelism"
"TODO: sdpa might be a good candidate for us to explore decomposed sharding propagation as it involves: matmul, pointwise, reduction ops together."

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 3]
```diff
@@ -60,6 +60,7 @@
 
 pointwise_ops = [
     # please keep the entries below alphabetically sorted
+    aten._conj.default,
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/pointwise_ops.py:63]: consider making this a separate PR with proper testing?

**Review by wanchaol — APPROVED** [review]
looks great!!

**Pull request closed without merging**